tensorflow/contrib/rnn/kernels/lstm_ops.h

```diff
@@ -0,0 +1,310 @@
+    return {batch_size_, cell_size_};
+  }
+
+ protected:
+  const int batch_size_;
+  const int input_size_;
+  const int cell_size_;
+};
+
+template <typename Device, typename T, bool USE_CUBLAS>
+struct LSTMCellBlockFprop : public LSTMCellBlock {
+  LSTMCellBlockFprop(const int batch_size, const int input_size,
+                     const int cell_size)
+    : LSTMCellBlock(batch_size, input_size, cell_size) {}
+
+  void operator()(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
+      const Device& d, const T forget_bias, typename TTypes<T>::ConstMatrix x,
+      typename TTypes<T>::ConstMatrix states_prev,
+      typename TTypes<T>::ConstMatrix w, typename TTypes<T>::ConstVec b,
+      typename TTypes<T>::Matrix cs_prev, typename TTypes<T>::Matrix h_prev,
+      typename TTypes<T>::Matrix xh, typename TTypes<T>::Matrix i,
+      typename TTypes<T>::Matrix cs, typename TTypes<T>::Matrix f,
+      typename TTypes<T>::Matrix o, typename TTypes<T>::Matrix ci,
+      typename TTypes<T>::Matrix co, typename TTypes<T>::Matrix icfo,
+      typename TTypes<T>::Matrix states, typename TTypes<T>::Matrix h) {
+    // [cs, h] = states_prev
+    cs_prev.device(d) =
+        states_prev.slice(states_cs_offsets(), cell_extents());
+
+    h_prev.device(d) =
+        states_prev.slice(states_h_offsets(), cell_extents());
+
+    // Concat xh = [x, h].
+    xh.slice(xh_x_offsets(), xh_x_extents()).device(d) = x;
+    xh.slice(xh_h_offsets(), xh_h_extents()).device(d) = h_prev;
+
+    // states1 = xh * w + b
+    typename TTypes<T>::ConstMatrix const_xh(xh.data(), xh.dimensions());
+    TensorBlasGemm<Device, T, USE_CUBLAS>::compute(
+        ctx, stream, d, false, false, T(1), const_xh, w, T(0), icfo);
+    icfo.device(d) +=
+        b.broadcast(Eigen::array<int, 2>({batch_size_, 1}));
+
+    // Input gate.
+    i.device(d) = icfo.slice(icfo_i_offsets(), cell_extents()).sigmoid();
+
+    // Cell input.
+    ci.device(d) = icfo.slice(icfo_c_offsets(), cell_extents()).tanh();
+
+    // Forget gate (w/ bias).
+    f.device(d) =
+        (icfo.slice(icfo_f_offsets(), cell_extents()) + f.constant(forget_bias))
+        .sigmoid();
+
+    // cs = ci .* i + f .* cs_prev
+    cs.device(d) = i * ci + f * cs_prev;
+
+    // co = tanh(cs)
+    co.device(d) = cs.tanh();
+
+    // Output gate.
+    o.device(d) = icfo.slice(icfo_o_offsets(), cell_extents()).sigmoid();
+
+    // h = o .* co
+    h.device(d) = o * co;
+
+    states.slice(states_cs_offsets(), cell_extents()).device(d) = cs;
+    states.slice(states_h_offsets(), cell_extents()).device(d) = h;
+  }
+};
+
+template <typename Device, typename T, bool USE_CUBLAS>
+struct LSTMCellBlockBprop : public LSTMCellBlock {
+  LSTMCellBlockBprop(const int batch_size, const int input_size,
+                     const int cell_size)
+    : LSTMCellBlock(batch_size, input_size, cell_size) {}
+
+  void operator()(
+      OpKernelContext* ctx, perftools::gputools::Stream* stream,
```

Comment:
Yeah I don't love this but suppose it will do for now.
On May 11, 2016 5:52 PM, "William Chan" <notifications@github.com> wrote:

> In tensorflow/contrib/rnn/kernels/lstm_ops.h
> <https://github.com/tensorflow/tensorflow/pull/2002#discussion_r62951156>:
>
> > +    // h = o .* co
> > +    h.device(d) = o * co;
> > +
> > +    states.slice(states_cs_offsets(), cell_extents()).device(d) = cs;
> > +    states.slice(states_h_offsets(), cell_extents()).device(d) = h;
> > +  }
> > +};
> > +
> > +template <typename Device, typename T, bool USE_CUBLAS>
> > +struct LSTMCellBlockBprop : public LSTMCellBlock {
> > +  LSTMCellBlockBprop(const int batch_size, const int input_size,
> > +                     const int cell_size)
> > +    : LSTMCellBlock(batch_size, input_size, cell_size) {}
> > +
> > +  void operator()(
> > +      OpKernelContext* ctx, perftools::gputools::Stream* stream,
>
> doing a grep of TTypes::ConstMatrix or TTypes::Matrix it seems like most
> other kernel functions don't use references or pointers for out params -- I
> think we should keep it as is?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2002/files/fc40971bba82762fe413b5e7d6fe12a09722876b#r62951156>
>